backend/clustering.py

```
@@ -0,0 +1,247 @@
+    """Cosine similarity for L2-normalized vectors."""
+    return float(np.dot(a, b))
+
+
+# ---------------------------------------------------------------------------
+# Clustering strategies
+# ---------------------------------------------------------------------------
+def cluster_agglomerative(embeddings: np.ndarray, sim_threshold: float = DEFAULT_SIM_THRESHOLD) -> np.ndarray:
+    """
+    Agglomerative clustering (average linkage, cosine distance threshold).
+
+    Time: ~O(n^2)–O(n^3); Memory: ~O(n^2); n = items.
```

Ruff (0.14.8)

Docstring contains ambiguous `–` (EN DASH). Did you mean `-` (HYPHEN-MINUS)?

(RUF002)